man/man1/oci-image-migrate-import.1
.\" Process this file with
.\" groff -man -Tascii oci-image-migrate-import.1
.\"
.\" Copyright (c) 2020 Oracle and/or its affiliates. All rights reserved.
.\" Licensed under the Universal Permissive License v 1.0 as shown
.\" at http://oss.oracle.com/licenses/upl.
.\"
.TH OCI-IMAGE-MIGRATE 1 "JAN 2020" Linux "User Manuals"
.SH NAME
oci-image-migrate-import \- assist the import of an image in the 
.B object storage
to the 
.B custom images
repository of the Oracle Cloud Infrastructure
.SH SYNOPSIS
.B oci-image-migrate-import  [-i|--image-name
.I image name
.B ] [-b|--bucket-name
.I bucket_name
.B ] [-c|--compartment-name
.I compartment name
.B ] [-d|--display-name
.I display name
.B ] [-l|--launch-mode
.I [PARAVIRTUALIZED|EMULATED|NATIVE]
.B ] [-y|--yes] [-v|--verbose] [--help]

.SH DESCRIPTION
The
.B oci-image-migrate-import
utility imports an image file residing in an object storage into the custom 
images repository of
.B Oracle Cloud Infrastructure.

.B oci-image-migrate-import
verifies the existence of the image file and the object storage.
It also checks if and image with the same name already exists
in the object storage.

.B oci-image-migrate-import
does not need to be run with root priviliges as long as the user has the
correct priviliges to access the
.B Oracle Cloud Infrastructure
and write priviliges in the directory where log and result files are written,
as well as in the log file itself (by default /var/tmp/oci-utils.log).

.B oci-image-migrate-import
requires the
.B oci-cli
utility installed and configured.

See
.BR https://docs.cloud.oracle.com/iaas/Content/API/SDKDocs/cliinstall.htm
or
.BR https://docs.cloud.oracle.com/iaas/Content/API/SDKDocs/climanualinst.htm
for more information.

.SH OPTIONS
.IP "-i|--image-name"
The name of the on-premise image. This option is mandatory.
.IP "-b|--bucket-name"
The name of the object storage or bucket in the
.B Oracle Cloud Infrastructure
where the imagefile is to be stored. This option is mandatory.
.IP "-c|--compartment-name"
The name of the destination compartment. This option is mandatory.
.IP "-d|--display-name"
The image name or the name the image will be saved in the
.B custom images
repository in the
.B Oracle Cloud Infrastructure.
This argument is optional, the default is the image name.
.IP "-l|--launch-mode"
The mode an instance created form the custom image will be started. Possible
modes are PARAVIRTUALIZED, EMULATED and NATIVE. The default is PARAVIRTUALIZED.
.IP "-y|--yes"
The answer on Yes/No questions is supposed to be yes
.IP "-v|--verbose"
Show more information about the import on the console.
.IP "--help"
Print a summary of the command line options.

.SH DIAGNOSTICS
Returns an exit status of 0 for success or 1 if an error occured.

.SH EXAMPLES
.PP
.nf
.RS
oci-image-migrate-import --image-name /images/OracleLinux7.vmdk -b MyObjectStorage -C MyCompartment -l EMULATED --display-name OracleLinux7U7 -v
.RE
.fi
.PP
Imports the image file /images/OracleLinux7.vmdk, on condition it exists, to the object storage of
.B Oracle Cloud Infrastructure
as OracleLinux7U7 if there is not yet and image with the same name present.

.SH SEE ALSO
.BR oci-image-migrate(1)
.BR oci-image-migrate-upload(1)
.BR ocid (8)
.BR oci-metadata (1)


.SH COPYRIGHT
Copyright (c) 2020 Oracle and/or its affiliates. All rights reserved.


NEWS
Version 0.11.0
- Added oci-image-migrate utility

Version 0.10.2
- Update to use Python 3 on OL8

Version 0.10.1
- Added support of libvirt network for KVM guests
- Added support for libvir storage pools

Version 0.9.1
- Added support for Oracle Linux 8 Beta 1

Version 0.9.0
- LINUX-498 - Enhance oci-metadata  for keys to only return key value 
              Added suport --value-only
- LINUX-560 - Cleanup utility not preserving permissions/ownerships (fixed)
              same as bug-29260959.
Version 0.8.0

- OLOCITOOLS-10 - added support for updating instance metadata for a specified instance.
                  $ oci-metadata [ -u|--update ] <key>=<value> ...
- OLOCITOOLS-11 - implemented method for retrieving metadata for other compute instances.
                  $ oci-metadata --instance-id <ocid>
- OLOCITOOLS-12 - implemented oci-metadata --export
                  $ oci-metadata --export -g key ..
                - enhanced metadata get by supporting path and wildcard in path.
                  $ oci-metadata -g /path/to/key -g /path/*/key
- OLOCITOOLS-28 - implemented oci-metadata --trim -g   ...
                  trim the path to the last key.
- OLOCITOOLS-24 - removed the metadata cach

- OLOCITOOLS-27 - fixed misleading error message when metadata does not include ssh_authorized_keys.

- fixed bug-28048699 - Needs to handle multiple physical NICs when creating secondary vnics
                       Added option "--nic-index" "to specify which physical NIC the vnic is assigned to.
- fixed bug-28599902 - oci-public-ip returns any public ip it found first.
                       Added an option [-a |--all] to list all public ips for this instance.
- fixed bug-28643343 - fixed most exceptions for oci config error.
- fixed bug-29026433 - oci-iscsi-config don't config secondary ip for second vnic automatically
                       when running delete ip option
- fixed bug-28840370 - local disks  should not count for "max_volumes" when creating iscsi disk
- fixed bug-29026422 - oci-iscsi-config -s " don't list boot partition for VM
- fixed bug-29118662 - oci-network-config detach vnic give wrong message

Version 0.7.1:

- fixed bug 28668447 - ocid needs to allow time for iSCSI connection to recover

Version 0.7:

- bump version for 0.7
- added oci-network-inspector to listing networking information.(OLOCITOOLS-5)
- added oci-growfs to support to grow filesystems (OLOCITOOLS-7)
  Currently only / boot volume is expendable.
- fixed bugs in oci-iscsi-config:
  1. bug-28493320: mis-list sdaa.. as partition of sda. It should be a seperate volume.
  2. bug 28615235 - different behavior for max_volumes setting for 32 or less than 32
  3. bug-28513898 - should not create a volume if it failed to attach                
  4. privilege adjustment for create and destroy(root+oci), attach and detach(root) 
  5. Bug 27835953 - objects can't be shown due to pagination issue.              
  6. Bug 28612818 - A detached volume can't be reattached to the instance saying max limit reached 
- fixed bug oci-network-config
  1. bug-28498139 - secondary ips on secondary VNICs not reachable issue.        
  2. Bug 28080064 - oci-network-config -s can't list all secondary private ip    
  3. Bug 28188853 - oci-network-config report error message when detach vnic     
- expanded 'OCI' to 'Oracle Cloud Infrastructure' in man pages, specfiles (OLOCITOOLS-8)

Version 0.6:

- bump version for 0.6
- add support for API access through Instance Principals
- add support for root using a designated user's OCI config files and keys
- oci_utils API automatically detects authentication method to be used
- ocid can discover secondary IP addresses and CHAP user/password using
  OCI API calls, if the Python SDK is configured or Instance Principals used
- network proxy support for making SDK calls
- configuration files for ocid: /etc/oci-utils.d/*
- support configuring the various functions of ocid individually,
  including refresh frequency or turning them off completely.
- ocid saves state and restores all volumes and VNIC configuration after reboot
- oci-network-config: new option: --detach-vnic
- oci-iscsi-config: new option: --destroy-volume
- oci-utils APIs are now thread safe
- NEW tool: oci-image-cleanup - a script that runs a set of cleanup steps to
  prepare the instance for snapshot
- oci-kvm utility rejects attempts to create guests if the required
  virtualization support is not enabled in the image it is being executed on


lib/oci_utils/impl/oci-image-migrate-import.py
# oci-utils
#
# Copyright (c) 2019, 2020 Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown
# at http://oss.oracle.com/licenses/upl.

"""
Script to import an on-premise virtual server from an object storage to the custom images 
in the Oracle Cloud Infrastructure. An image which passed the checks and modifications 
of - and was uploaded by the oci-image-migrate utility should import fine.
"""
import argparse
import json
import logging.config
import six
import sys
import time

from oci_utils.migrate import console_msg, exit_with_msg, read_yn
from oci_utils.migrate import migrate_tools
from oci_utils.migrate import migrate_utils

_logger = logging.getLogger('oci-utils.import_ci')


def parse_args():
    """
    Parse the command line arguments and return an object representing the command
    line as returned by the argparse's parse-args().
    arguments:
    -i|--image-name <image name>; mandatory.
    -b|--bucket-name <bucket name>; mandatdory.
    -c|--compartment-name <compartment name>; mandatory.
    -l|--launch-mode [PARAVIRTUALIZED|EMULATED|NATIVE]; optional, the default
         is PARAVIRTUALIZED.
    -d|--displayname <display name>; optional.
    
    Returns
    -------
        The command line namespace.
    """
    parser = argparse.ArgumentParser(
        description='Utility to import a (verified and modified) on-premise '
                    'legacy images which was uploaded to object storage in '
                    'the custom images folderof the Oracle Cloud Infrastructure.',
        add_help=False)
    #
    parser.add_argument('-i', '--image-name',
                        action='store',
                        dest='imagename',
                        required=True,
                        help='The name of the object representing the uploaded image.')
    parser.add_argument('-b', '--bucket-name',
                        action='store',
                        dest='bucketname',
                        required=True,
                        help='The name of the object storage.')
    parser.add_argument('-c', '--compartment-name',
                        action='store',
                        dest='compartmentname',
                        required=True,
                        help='The name of the destination compartment.')
    parser.add_argument('-d', '--display-name',
                        action='store',
                        dest='displayname',
                        help='Image name as it will show up in the custom '
                             'images; the default is the image name.',
                        default=None)
    parser.add_argument('-l', '--launch-mode',
                        action='store',
                        dest='launchmode',
                        help='The mode the instance created from the custom '
                             'image will be started; the default '
                             'is PARAVIRTUALIZED.',
                        choices=['PARAVIRTUALIZED', 'EMULATED', 'NATIVE'],
                        default='PARAVIRTUALIZED')
    parser.add_argument('--help', action='help', help='Display this help')
    parser._optionals.title = 'Arguments'
    args = parser.parse_args()
    if args.displayname is None:
        args.displayname = args.imagename
    return args

def main():
    #
    # command line
    cmdlineargs = parse_args()
    console_msg(msg = 'Importing %s from %s into %s as %s and setting '
                      'launchmode to %s.' % (cmdlineargs.imagename,
                                             cmdlineargs.bucketname,
                                             cmdlineargs.bucketname,
                                             cmdlineargs.displayname,
                                             cmdlineargs.launchmode))
    compartment = cmdlineargs.compartmentname
    bucket = cmdlineargs.bucketname
    objectname = cmdlineargs.imagename
    displayname = cmdlineargs.displayname
    launchmode = cmdlineargs.launchmode
    #
    # oci config file 
    try:
        ocicfg = migrate_utils.get_oci_config()
        console_msg('Found oci config file')
        tenancy = ocicfg['tenancy']
        console_msg(msg = 'tenancy = %s' % tenancy)
        _logger.debug('Tenancy: %s' % tenancy)
        cmd = ['oci', 'iam', 'compartment', 'list', '-c', '%s' % tenancy, '--all']
        console_msg(msg='running: %s' % cmd)
        cmpdict = json.loads(migrate_tools.run_popen_cmd(cmd))
    except Exception as e:
        exit_with_msg('Failed to locate or read the oci config file, is oci '
                      'cli installed?: %s' % str(e))
    #
    # namespace
    cmd = ['oci', 'os', 'ns', 'get']
    try:
        nsdict = json.loads(migrate_tools.run_popen_cmd(cmd))
        namespace = nsdict['data']
        console_msg(msg='namespace: %s' % namespace)
    except Exception as e:
        exit_with_message('Failed to identify the namespace: %s' % str(e))
    #
    # compartment
    try:
        foundcom = False
        for k,v in cmpdict.items():
            for x in v:
                if x['name'] == compartment:
                    compdata = x
                    # print(compdata)
                    console_msg(msg='compartment data: %s' % compdata['id'])
                    compartment_id = compdata['id']
                    foundcom = True
                    break
        if not foundcom:
            exit_with_msg('Failed to find %s' % compartment)
    except Exception as e:
        exit_with_msg('Failed to find %s: %s' % (compartment, str(e)))
    #
    # bucket
    try:
        bucketcontents = migrate_utils.bucket_exists(bucket)
        console_msg(msg='Object storage %s present.' % bucket)
    except Exception as e:
        exit_with_msg('Object storage %s not found: %s' % (bucket, str(e)))
    #
    # bucket data
    try:
        cmd = ['oci', 'os', 'bucket', 'get', '--bucket-name', '%s' % bucket]
        console_msg(msg='running: %s' % cmd)
        bucketlist = migrate_tools.run_popen_cmd(cmd)
        if bucketlist is not None:
            bucketdict = json.loads(bucketlist)
        else:
            console_msg(msg='bucket data for %s is empty.' % bucket)
    except Exception as e:
        exit_with_msg('Failed to get data of %s: %s' % (bucket, str(e)))
    #
    # bucket contents
    try:
        cmd = ['oci', 'os', 'object', 'list', '--bucket-name', '%s' %  bucket]
        console_msg('running: %s' % cmd)
        bucketcontents = migrate_tools.run_popen_cmd(cmd)
        if bucketcontents is not None:
            bucketcontentsdict = json.loads(bucketcontents)
        else:
            console_msg(msg='bucket %s is emtpy.' % bucket)
        #print('%s' % bucketcontentsdict)
    except Exception as e:
        exit_with_msg('Failed to get contents of %s: %s' % (bucket, str(e)))    
    #
    # object exists?
    if migrate_utils.object_exists(bucketcontents, objectname):
        console_msg(msg='Object %s present in %s.' % (objectname, bucket))
    else:
        exit_with_msg('Object %s missing from %s.' % (objectname, bucket))
    #
    # display name exists?
    cmd = ['oci', 'compute', 'image', 'list', '--compartment-id',
           '%s' % compartment_id, '--display-name',
           '%s' % displayname]
    objstat = migrate_tools.run_popen_cmd(cmd)
    if objstat is not None:
        exit_with_msg('Image with name %s already exists.' % displayname)
    #
    # upload?
    if not read_yn('\n  Uploading %s to %s as %s'
                   % (objectname, compartment, displayname)):
        exit_with_msg('Exiting.\n')
    cmd = ['oci', 'compute', 'image', 'import', 'from-object', '--bucket-name',
           '%s' % bucket, '--compartment-id', '%s' % compartment_id,
           '--name', '%s' % objectname, '--namespace', '%s' % namespace,
           '--launch-mode',  launchmode, '--display-name', '%s' % displayname]
    console_msg(msg='running: %s' % cmd)
    try:
        resval = migrate_tools.run_popen_cmd(cmd)
        #console_msg(msg='running import: %s' % resval)
    except Exception as e:
        exit_with_msg('Failed to import %s: %s' % (objectname, str(e)))
    #
    # follow up
    cmd = ['oci', 'compute', 'image', 'list', '--compartment-id',
           '%s' % compartment_id, '--display-name', '%s' % displayname]
    # console_msg(msg='running: %s' % cmd)
    finished = False
    year, month, day, hour, min = map(int, time.strftime("%Y %m %d %H %M").split())
    sys.stdout.write('\n  %02d-%02d-%04d %02d:%02d %s\n'
                     % (day, month, year, hour, min,
                        'Starting import of %s' % displayname))
    sys.stdout.flush()
    cnt=0
    modcnt=40
    addn = -1
    while not finished:
        objstat = json.loads(migrate_tools.run_popen_cmd(cmd))
        for ob in objstat['data']:
            if ob['display-name'] == displayname:
                if ob['lifecycle-state'] == 'AVAILABLE':
                    finished = True
                else:
                    addn *= -1
                    for i in range(modcnt-1):
                        cntstr = (cnt%modcnt+1)*'.' + (modcnt-cnt%modcnt)*' '
                        year, month, day, hour, min \
                            = map(int, time.strftime("%Y %m %d %H %M").split())
                        sys.stdout.write('  %02d-%02d-%04d %02d:%02d %s %s\r'
                                         % (day, month, year, hour, min,
                                            'Still importing %s' % displayname, cntstr))
                        sys.stdout.flush()
                        time.sleep(1)
                        cnt += 1
            else:
                console_msg(msg='%s not found.' % displayname)
    console_msg(msg='Done')

    return(0)


if __name__ == "__main__":
    sys.exit(main())
